Conditional Wx3 import

devos50 opened the pull request:
This PR contains some fixes for Wx3.

The most important change is an conditional check in the `tribler.py` file that checks whether Wx3 is installed on the users' system. If so, it uses Wx3. If not, it fallbacks to Wx2.8. In some sense, we now have an experimental version of Wx3 in Tribler.

I've tested manually on Windows, Ubuntu and OS X and everything still seems to work ok. Wx3 on OS X is quite stable currently with some minor issues left.

whirm commented:
@devos50 how long do you think it would take to have the OSX issues fixed? 
Do you know if they where there there before the migration?
Do you have a list of them?

devos50 commented:
I'm not sure. There's an issue with the tooltip that keeps visible when moving the cursor to other elements. Moreover, clicking elements in a list is a bit buggy but I have no idea where this comes from.

devos50 commented:
Oh that dispersy submodule change should probably not be there, let me change that.

synctext commented:
hard coded /usr/bin/python :)
after 1million new users, use the sysvariable..

whirm commented:
@synctext we really want to use the system's python interpreter, not any other that the user may have installed by itself and added to the PATH.

devos50 commented:
@whirm I fixed the Dispersy submodule pointer, the submodule was in a detached head state apparently but I fixed by checking out to the most recent commit.

whirm commented:
@devos50 it's always in a detached head state when you update a submodule (you normally refer to a specific commit, not a branch)

whirm commented:
Have you tested these changes with wx-2.8 to make sure we are still backwards compatible?

I'm not saying we should be if it means any more effort on our part. If there's any new issue at all, let's just stick to 3.0 from now on.

devos50 commented:
Yes, I've tested with wx-2.8 also.

whirm commented:
I just started your branch on my Debian SID/amd64 and it crashes right away:
```
└──> ./tribler.sh

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed
Traceback (most recent call last):
  File "/var/data/projects/delft/tribler/Tribler/Main/tribler_main.py", line 225, in __init__
    self.splash.tick)
  File "/var/data/projects/delft/tribler/Tribler/Main/vwxGUI/MainFrame.py", line 202, in __init__
    self.channeldetailspanel = ChannelDetails(self.splitter_bottom_window)
  File "/var/data/projects/delft/tribler/Tribler/Main/vwxGUI/list_details.py", line 1466, in __init__
    self.createAllTabs()
  File "/var/data/projects/delft/tribler/Tribler/Main/vwxGUI/list_details.py", line 1496, in createAllTabs
    self.detailsTab.Layout()
  File "/usr/lib/python2.7/dist-packages/wx-3.0-gtk2/wx/_core.py", line 11496, in Layout
    return _core_.Window_Layout(*args, **kwargs)
PyAssertionError: C++ assertion "m_growableRows[n] < nrows" failed at ../src/common/sizer.cpp(1874) in AdjustForGrowables(): invalid growable row index

(tribler.py:3072): GLib-GObject-WARNING **: invalid (NULL) pointer instance

(tribler.py:3072): GLib-GObject-CRITICAL **: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed

(tribler.py:3072): Gtk-CRITICAL **: IA__gtk_window_set_modal: assertion 'GTK_IS_WINDOW (window)' failed
^C

```
I'm running:
```
ii  python-wxgtk3.0                   3.0.2.0+dfsg-1+b1     amd64                 Python interface to the wxWidgets Cross-platform C++ GUI toolkit
```

devos50 commented:
@whirm saw this error various times when I started analysing Wx3 compatibility. I removed the `addGrowableRow` call for two reasons, the first one being that the index is incorrect and the second one is the fact that we probably do not need a growable index in here.

Could you test my branch again and let me know whether it works?

whirm commented:
Seems to be working fine now, but I still get these assertions:

```
(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed

(tribler.py:19470): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion 'width >= -1' failed
```

devos50 commented:
retest this please

whirm commented:
retest this please

devos50 commented:
@whirm can this one be merged? After this one is merged, it is possible for me to work on OS X (since I cannot run wx2.8 on OS X).

whirm commented:
Are the assertions fixed?

devos50 commented:
@whirm they are not critical, see http://anonscm.debian.org/cgit/collab-maint/wx-migration-tools.git/tree/README, especially the part about Wx3 assertions. I can suppress them (like in wx2.8 by default) if you want?

whirm commented:
@devos50 do they get into the `Tribler.exe.log` file on windows? If so, it means Tribler will report errors happened when shutting down every single time.

I'll merge this and create an issue for the assertion.

whirm commented:
Oh, but first join the commits by topic (as in commits fixing stuff you broke on a previous commit)

devos50 commented:
@whirm done

devos50 commented:
Thanks, now I can also start running the unit tests on OS X!
